### PR TITLE
extend ILDAPProvider to allow reading arbitrairy ldap attributes for users

### PR DESCRIPTION
**Pull request by icewind1991**
For allowing things like #24955 using public interfaces

### CODE REVIEW

File: lib/public/LDAP/ILDAPProvider.php [hunk 1]
```diff
@@ -157,4 +157,14 @@ public function getLDAPEmailField($uid);
...
+
+	/**
+	 * Get an LDAP attribute for a nextcloud user
+	 * @param string $uid the nextcloud user id to get the attribute for
+	 * @param string $attribute the name of the attribute to read
+	 * @return string|null
+	 * @throws \Exception if user id was not found in LDAP
+	 * @since 22.0.0
+	 */
+	public function getLDAPAttribute(string $uid, string $attribute): ?string;
```
Comment (blizzz) [inline lib/public/LDAP/ILDAPProvider.php:169]: call it `getUserAttribute`, please, to underline it's not about groups. The LDAP part should be clear through context.

File: apps/user_ldap/lib/LDAPProvider.php [hunk 2]
```diff
@@ -306,4 +306,35 @@ public function getLDAPGroupMemberAssoc($gid) {
...
+		$connection = $access->getConnection();
+		$key = $uid . "::" . $attribute;
+		$cached = $connection->getFromCache($key);
+
+		if ($cached !== null) {
+			return $cached;
+		}
+
+		$value = $access->readAttribute($access->username2dn($uid), $attribute);
+		if (count($value) > 0) {
```
Comment (blizzz) [inline apps/user_ldap/lib/LDAPProvider.php:331]:
```suggestion
		if (is_array($value) && count($value) > 0) {
```

`$value` may be  `false`

**Review by blizzz — CHANGES_REQUESTED** [review]

**Comment by icewind1991** [comment]
@blizzz done

File: lib/public/LDAP/ILDAPProvider.php [hunk 3]
```diff
@@ -157,4 +157,14 @@ public function getLDAPEmailField($uid);
...
 	 */
 	public function getLDAPGroupMemberAssoc($gid);
+
+	/**
+	 * Get an LDAP attribute for a nextcloud user
+	 * @param string $uid the nextcloud user id to get the attribute for
+	 * @param string $attribute the name of the attribute to read
+	 * @return string|null
+	 * @throws \Exception if user id was not found in LDAP
+	 * @since 22.0.0
```
Comment (MorrisJobke) [inline lib/public/LDAP/ILDAPProvider.php:167]:
```suggestion
	 * @since 21.0.0
```

Let's do it for 21.

**Review by MorrisJobke — APPROVED** [review]
Code looks good, but some tests would be nice.

**Review by blizzz — APPROVED** [review]

**Comment by MorrisJobke** [comment]
> For allowing things like #24955 using public interfaces

Mind to submit this with the new API?

**Comment by MorrisJobke** [comment]
> > 
> 
> Mind to submit this with the new API?

cc @blizzz @icewind1991

**Comment by MorrisJobke** [comment]
> > For allowing things like #24955 using public interfaces
> 
> Mind to submit this with the new API?

@blizzz @icewind1991 Any news here?

**Comment by blizzz** [comment]
Last was https://github.com/nextcloud/server/pull/25010#issuecomment-756114527

**Comment by icewind1991** [comment]
https://github.com/nextcloud/server/pull/26397

**Pull request merged**